kolibri/core/content/utils/content_request.py

```diff
@@ -352,24 +435,53 @@ def process_metadata_import(incomplete_downloads_without_metadata):
     :param incomplete_downloads_without_metadata: a ContentDownloadRequest queryset
     :type incomplete_downloads_without_metadata: django.db.models.QuerySet
     """
-    peer = _get_preferred_network_location(version_filter=">=0.16.0")
-    if not peer:
-        # can't import metadata without peers having minimum version 0.16.0
-        logger.info("No acceptable peer network device for importing content metadata")
-        return
+    source_instance_ids = list(
+        incomplete_downloads_without_metadata.values_list(
+            "source_instance_id", flat=True
+        ).distinct()
+    )
+    # append `None` which will bypass the explicit peer selection and use a trusted peer
+    # of a recent sync, if available, during the last step of the process
+    source_instance_ids.append(None)
+    peers = []
+
+    for source_instance_id in source_instance_ids:
+        # get the peer for this source instance
+        peer = _get_preferred_network_location(
+            instance_id=source_instance_id, version_filter=">=0.16.0"
+        )
+        if not peer:
+            continue
 
-    # during processing, if there's a critical failure in making requests to the peer, this will
-    # capture those errors, and obviously the raise exceptions will interrupt processing
-    with capture_connection_state(peer):
-        with NetworkClient.build_from_network_location(peer) as client:
-            # test connection
-            client.connect()
-            _import_metadata(
-                client,
-                incomplete_downloads_without_metadata.values_list(
-                    "contentnode_id", flat=True
-                ),
+        # store all available peers, so we can fall back to them if later preferred peers fail
+        peers.append(peer)
+        filters = dict()
+        if source_instance_id is not None:
+            filters["source_instance_id"] = source_instance_id
+
+        # start with the most recent peer, and work backwards
+        for peer in reversed(peers):
+            # during processing, if there's a critical failure in making requests to the peer,
+            # this will capture those errors, and obviously the raising of exceptions
+            # will interrupt processing
+            with capture_connection_state(peer):
+                with NetworkClient.build_from_network_location(peer) as client:
+                    # test connection
+                    client.connect()
+                    _import_metadata(
```

Comment:
Could we return the number of successfully imported content nodes from this function, so we can break from the loop once we've finished importing everything?

Alternatively, we can evaluate the unprocessed count after each import instead of outside of the loop.

Reply:
Yes :+1: